File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class UpdateStrategy {
  update(_item: Item): Item {
    throw new Error("Not implemented");
  }
}

export class BrieUpdateStrategy extends UpdateStrategy {
  update(item: Item): Item {
    if (item.quality < 50) {
      item.quality = item.quality + 1;
    }
    item.sellIn = item.sellIn - 1;

    if (item.sellIn < 0 && item.quality < 50) {
      item.quality = item.quality + 1;
    }

    return item;
  }
}

export class BackstagePassesUpdateStrategy extends UpdateStrategy {
  changeRate(item: Item): number {
    if (item.sellIn < 6) {
      return 3;
    } else if (item.sellIn < 11) {
      return 2;
    }
    return 1;
  }

  update(item: Item): Item {
    const changeRate = this.changeRate(item);

    item.sellIn = item.sellIn - 1;

    if (item.sellIn < 0) {
      item.quality = 0;
      return item;
    }

    item.quality = Math.min(50, item.quality + changeRate);
    return item;
  }
}

export class LegendaryUpdateStrategy extends UpdateStrategy {
  update(item: Item): Item {
    return item;
  }
}

export class ConjuredUpdateStrategy extends UpdateStrategy {
  changeRate(item: Item): number {
    return item.sellIn < 0 ? -4 : -2;
  }

  update(item: Item): Item {
    item.sellIn = item.sellIn - 1;
    item.quality = Math.max(0, item.quality + this.changeRate(item));
    return item;
  }
}

export class NormalUpdateStrategy extends UpdateStrategy {
  changeRate(item: Item): number {
    return item.sellIn < 0 ? -2 : -1;
  }

  update(item: Item): Item {
    item.sellIn = item.sellIn - 1;
    item.quality = Math.max(0, item.quality + this.changeRate(item));
    return item;
  }
}

export class GildedRose {
  items: Array<Item>;
  strategies: { [key: string]: UpdateStrategy };

  constructor(items = [] as Array<Item>) {
    this.items = items;
    this.strategies = {
      "Aged Brie": new BrieUpdateStrategy(),
      "Backstage passes to a TAFKAL80ETC concert":
        new BackstagePassesUpdateStrategy(),
      "Sulfuras, Hand of Ragnaros": new LegendaryUpdateStrategy(),
      "Conjured Mana Cake": new ConjuredUpdateStrategy(),
    };
  }

  updateQuality(): Array<Item> {
    this.items.forEach((item) => {
      const strategy = this.strategies[item.name] || new NormalUpdateStrategy();
      strategy.update(item);
    });

    return this.items;
  }
}

File: ruby/gilded_rose.rb
class GildedRose

  MAX_QUALITY = 50
  MIN_QUALITY = 0
  QUALITY_CHANGE = 1
  SELL_IN_EXPIRED = 0
  SELL_IN_10_DAYS = 10
  SELL_IN_5_DAYS = 5


  def initialize(items)
    @items = items
  end
  
  def update_quality()
    @items.each do |item|
      case item.name
      when  "Aged Brie"
        update_aged_brie(item)
      when "Sulfuras, Hand of Ragnaros"
        update_legendary_item(item)
      when "Backstage passes to a TAFKAL80ETC concert"
        update_backstage_passes(item)
      else
        update_normal_item(item)
      end
    end
  end

  private

  def update_normal_item(item)
    if item.quality > MIN_QUALITY
      item.quality -= QUALITY_CHANGE
    end
    item.sell_in -= 1
    if item.sell_in < SELL_IN_EXPIRED
      if item.quality > MIN_QUALITY
        item.quality = MIN_QUALITY
      end
    end
  end

  def update_aged_brie(item)
    if item.quality < MAX_QUALITY
      item.quality += QUALITY_CHANGE
    end
    item.sell_in -= 1
    if item.sell_in < SELL_IN_EXPIRED
      if item.quality < MAX_QUALITY
        item.quality += QUALITY_CHANGE
      end
    end
  end

  def update_legendary_item(item)
    # do nothing
  end

  def update_backstage_passes(item)
    item.sell_in -= 1
    if item.sell_in < SELL_IN_EXPIRED
      item.quality = MIN_QUALITY
    else
      item.quality += QUALITY_CHANGE if item.quality < MAX_QUALITY
      item.quality += QUALITY_CHANGE if item.sell_in < SELL_IN_10_DAYS && item.quality < MAX_QUALITY
      item.quality += QUALITY_CHANGE if item.sell_in < SELL_IN_5_DAYS && item.quality < MAX_QUALITY
    end
  end
end

class Item
  attr_accessor :name, :sell_in, :quality

  def initialize(name, sell_in, quality)
    @name = name
    @sell_in = sell_in
    @quality = quality
  end

  def to_s()
    "#{@name}, #{@sell_in}, #{@quality}"
  end
end
